src/EditorFeatures/Test2/IntelliSense/CSharpCompletionCommandHandlerTests.vb

```diff
@@ -4045,8 +4097,111 @@ class C
+        <WpfTheory, Trait(Traits.Feature, Traits.Features.Completion)>
+        Public Async Function TestAfterCasePatternSwitchLabel(completionImplementation As CompletionImplementation) As Task
+            Using state = TestStateFactory.CreateCSharpTestState(completionImplementation,
+                              <Document>
+class C
+{
+    void M()
+    {    
+        object o = 1;
+        switch(o)
+        {
+            case int i:
+                $$
+                break;
+        }
+    }
+}
+                              </Document>)
+
+                state.SendTypeChars("this")
+                Await state.AssertSelectedCompletionItem(displayText:="this", isHardSelected:=True)
+            End Using
+        End Function
+
+        <InlineData(CompletionImplementation.Modern)> ' New test: fails in the legacy completion
+        <WpfTheory, Trait(Traits.Feature, Traits.Features.Completion)>
+        Public Async Function TestBackspaceInMiddleOfSelection(completionImplementation As CompletionImplementation) As Task
+            Using state = TestStateFactory.CreateCSharpTestState(completionImplementation,
+                              <Document>
+public enum foo
+{
+    aaa
+}
+
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        foo.a$$a
+    }
+}
+                              </Document>)
+
+                state.Workspace.Options = state.Workspace.Options.WithChangedOption(
+                    CompletionOptions.TriggerOnDeletion, LanguageNames.CSharp, True)
+
+                state.SendInvokeCompletionList()
+                state.SendBackspace()
+                Await state.AssertSelectedCompletionItem(displayText:="aaa", isSoftSelected:=True)
+            End Using
+        End Function
+
+        <MemberData(NameOf(AllCompletionImplementations))>
+        <WpfTheory, Trait(Traits.Feature, Traits.Features.Completion)>
+        Public Async Function TestBackspaceWithMultipleCharactersSelected(completionImplementation As CompletionImplementation) As Task
+            Using state = TestStateFactory.CreateCSharpTestState(completionImplementation,
+                              <Document>
+using System;
+
+public class Program
+{
+    public static void Main(string[] args)
+    {
+        Console.WriteLine$$
+    }
+}
+                              </Document>)
+
+                state.Workspace.Options = state.Workspace.Options.WithChangedOption(
+                    CompletionOptions.TriggerOnDeletion, LanguageNames.CSharp, True)
+
+                state.SendInvokeCompletionList()
+                state.SelectAndMoveCaret(-6)
+                state.SendBackspace()
+                Await state.AssertSelectedCompletionItem(displayText:="Write", isHardSelected:=True)
+            End Using
+        End Function
+
+        <InlineData(CompletionImplementation.Modern)> ' New test: improves the behavior comparing with the legacy implementation
```

Comment:
><InlineData(CompletionImplementation.Modern)> ' New test: improves the behavior comparing with the legacy implementation [](start = 8, length = 120)

@cyrusnajmabadi, this test contradicts TestDoNotInsertColonSoThatUserCanCompleteOutAVariableNameThatDoesNotCurrentlyExist_IE_TheCyrusCase with your name in the test name.

The desired behavior is to add a colon here. However, NamedParameterCompletionProvider intentionally removes the colon.

What do you think?

Reply:
Note: IMO we should absolutely *not* be changing behavior while *also* changing implementation.  If we want to change behavior, we should do that either fully before, or fully after, we change the implementation.  Doing both at the same time is going to lead to things changing that we did not intend, and which we likely will not discover or be able to change easily.

Reply:
As another example: we should be able to ship an A/B test where we run both systems at the same time and we NFW if the new system deviates from the old system in any way.  However, if behavior changes, we cannot do that.

Reply:
I'd prefer we just not change behavior in this PR *period* :)

If we would like to change things, it should be based on an issue filed with a strong argument as to why the existing behavior is not good, and why we'd think this was better.  :)   We can then schedule that behavioral change before/after this PR as appropriate.

Reply:
OK. Makes sense. It seems to be easy to isolate this change.

---
In reply to: [223562097](https://github.com/dotnet/roslyn/pull/29016#discussion_r223562097) [](ancestors = 223562097)